File: Go/unique_email_addresses/929.go
package uniqueemailaddresses

import (
	"strings"
)

func numUniqueEmails(emails []string) int {
	uniqueEmails := []string{}
	for _, email := range emails {
		part := strings.Split(email, "@")
		local_name := part[0]
		domain_name := part[1]
		local_name = strings.ReplaceAll(local_name, ".", "")
		if strings.Contains(local_name, "+") {
			local_name, _, _ = strings.Cut(local_name, "+")
		}
		cleanEmail := local_name + "@" + domain_name
		if !func(email string) bool {
			for _, v := range uniqueEmails {
				if v == email {
					return true
				}
			}
			return false
		}(cleanEmail) {
			uniqueEmails = append(uniqueEmails, cleanEmail)
		}
	}
	return len(uniqueEmails)
}

// Second solution
func numUniqueEmails2(emails []string) int {
	hashSet := map[string]bool{}
	for _, email := range emails {
		cleanEmail := strings.Builder{}
		part := strings.Split(email, "@")
		localName := part[0]
		domainName := part[1]
		for i := 0; i < len(localName) && localName[i] != '+'; i++ {
			if localName[i] == '.' {
				continue
			}
			cleanEmail.WriteString(string(localName[i]))
		}
		hashSet[cleanEmail.String()+"@"+domainName] = true
	}
	return len(hashSet)
}

File: TypeScript/unique_emails/929.ts
function numUniqueEmails(emails: string[]): number {
  const uniqueEmails: string[] = [];
  for (const e of emails) {
    const part = e.split("@");
    let localName = part[0];
    const domainName = part[1];

    localName = localName.replaceAll(".", "");
    if (localName.includes("+")) {
      localName = localName.substring(0, localName.indexOf("+"));
    }
    const cleanEmail = localName + "@" + domainName;
    if (!uniqueEmails.includes(cleanEmail)) {
      uniqueEmails.push(cleanEmail);
    }
  }
  return uniqueEmails.length;
}

// Second solution
function numUniqueEmails2(emails: string[]): number {
  const hashSet = new Set<string>();
  for (const email of emails) {
    let cleanEmail = "";
    const part = email.split("@");
    const localName = part[0];
    const domainName = part[1];
    for (let i = 0; i < localName.length && localName[i] !== "+"; ++i) {
      if (localName[i] === ".") {
        continue;
      }
      cleanEmail += localName[i];
    }
    hashSet.add(cleanEmail + "@" + domainName);
  }
  return hashSet.size;
}

export { numUniqueEmails, numUniqueEmails2 };
